core/templates/services/context.service.spec.ts

```diff
@@ -43,6 +43,11 @@ describe('Context service', () => {
       expect(ecs.getExplorationId()).toBe('123');
     });
 
+    it('should correctly retrieve the exploration id cached before', () => {
+      expect(ecs.getExplorationId()).toBe('123');
```

Comment:
Hm, to make it clearer that this test is actually checking that the cached value is returned. Can we have the first line expect that the id is '123'. Then change the URL params so that the URL now has a different exp id and then call `getExplorationId()` and expect that it is still '123'?

Reply:
I already tried it earlier and was thinking the same as you but its not possible.

without spyOn yet:- will throw error

spyOn('123'): will set explorationId to '123' and will return that.
Again setting spyOn('456'): will throw error (can't use spyOn for mocking twice in test)

Reply:
Yup you can't mock the same function twice in the same describe block. My suggestion was to actually update the URL (e.g. changing `windowRef.nativeWindow.location` or similar and have getPathname() actually execute). Would that work?

e.g.
https://github.com/oppia/oppia/blob/db8b32f3e8799f896bc16511cd1014b0df660488/core/templates/pages/landing-pages/topic-landing-page/topic-landing-page.component.spec.ts#L93-L98

Reply:
urlService.windowRef is private entity. Can't excess.

Reply:
In the example above, windowRef is a private variable (see corresponding component).

Reply:
Done!